File: modules/debugger/src/main/java/com/falsepattern/zigbrains/debugger/toolchain/ZigDebuggerToolchainService.java
package com.falsepattern.zigbrains.debugger.toolchain;

import com.falsepattern.zigbrains.ZigBundle;
import com.falsepattern.zigbrains.common.ZigPathManager;
import com.falsepattern.zigbrains.debugger.settings.MSVCDownloadPermission;
import com.falsepattern.zigbrains.debugger.settings.ZigDebuggerSettings;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.PathManager;
import com.intellij.openapi.components.Service;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogBuilder;
import com.intellij.openapi.util.SystemInfo;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.ui.BrowserHyperlinkListener;
import com.intellij.ui.HyperlinkLabel;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBPanel;
import com.intellij.util.download.DownloadableFileService;
import com.intellij.util.io.Decompressor;
import com.intellij.util.system.CpuArch;
import com.intellij.util.system.OS;
import com.jetbrains.cidr.execution.debugger.CidrDebuggerPathManager;
import com.jetbrains.cidr.execution.debugger.backend.bin.UrlProvider;
import com.jetbrains.cidr.execution.debugger.backend.lldb.LLDBDriverConfiguration;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import lombok.val;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.Properties;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicReference;

@Service(Service.Level.APP)
public final class ZigDebuggerToolchainService {
    private MSVCMetadataProvider msvcProvider = new MSVCMetadataProvider();
    public static ZigDebuggerToolchainService getInstance() {
        return ApplicationManager.getApplication().getService(ZigDebuggerToolchainService.class);
    }

    public DebuggerAvailability debuggerAvailability(DebuggerKind kind) {
        return switch (kind) {
            case LLDB -> lldbAvailability();
            case GDB -> gdbAvailability();
            case MSVC -> msvcAvailability();
        };
    }

    public DebuggerAvailability lldbAvailability() {
        if (LLDBDriverConfiguration.hasBundledLLDB())
            return DebuggerAvailability.Bundled;

        final String frameworkPath, frontendPath;
        if (SystemInfo.isMac) {
            frameworkPath = "LLDB.framework";
            frontendPath = "LLDBFrontend";
        } else if (SystemInfo.isUnix) {
            frameworkPath = "lib/liblldb.so";
            frontendPath = "bin/LLDBFrontend";
        } else if (SystemInfo.isWindows) {
            return DebuggerAvailability.Unavailable;
//            frameworkPath = "bin/liblldb.dll";
//            frontendPath = "bin/LLDBFrontend.exe";
        } else {
            return DebuggerAvailability.Unavailable;
        }

        val lldbPath = lldbPath();
        val frameworkFile = lldbPath.resolve(frameworkPath);
        val frontendFile = lldbPath.resolve(frontendPath);
        if (!Files.exists(frameworkFile) || !Files.exists(frontendFile))
            return DebuggerAvailability.NeedToDownload;

        val versions = loadDebuggerVersions(DebuggerKind.LLDB);
        val urls = lldbUrls();
        if (urls == null)
            return DebuggerAvailability.Unavailable;

        val lldbFrameworkVersion = fileNameWithoutExtension(urls.framework().toString());
        val lldbFrontendVersion = fileNameWithoutExtension(urls.frontend().toString());

        if (!Objects.equals(versions.get(LLDB_FRAMEWORK_PROPERTY_NAME), lldbFrameworkVersion) ||
            !Objects.equals(versions.get(LLDB_FRONTEND_PROPERTY_NAME), lldbFrontendVersion))
            return DebuggerAvailability.NeedToUpdate;

        return new DebuggerAvailability.LLDBBinaries(frameworkFile, frontendFile);
    }

    public DebuggerAvailability gdbAvailability() {
        // Even if we have bundled GDB, it still doesn't work on macOS for local runs
        if (SystemInfo.isMac)
            return DebuggerAvailability.Unavailable;

        if (CidrDebuggerPathManager.getBundledGDBBinary().exists())
            return DebuggerAvailability.Bundled;

        final String gdbBinaryPath;
        if (SystemInfo.isUnix) {
            gdbBinaryPath = "bin/gdb";
        } else if (SystemInfo.isWindows) {
            return DebuggerAvailability.Unavailable;
//            gdbBinaryPath = "bin/gdb.exe";
        } else {
            return DebuggerAvailability.Unavailable;
        }

        val gdbFile = gdbPath().resolve(gdbBinaryPath);
        if (!Files.exists(gdbFile))
            return DebuggerAvailability.NeedToDownload;

        val versions = loadDebuggerVersions(DebuggerKind.GDB);
        val gdbUrl = gdbUrl();
        if (gdbUrl == null)
            return DebuggerAvailability.Unavailable;

        val gdbVersion = fileNameWithoutExtension(gdbUrl.toString());

        if (!Objects.equals(versions.get(GDB_PROPERTY_NAME), gdbVersion))
            return DebuggerAvailability.NeedToUpdate;

        return new DebuggerAvailability.GDBBinaries(gdbFile);
    }

    public DebuggerAvailability msvcAvailability() {
        //Only applies to Windows
        if (!SystemInfo.isWindows)
            return DebuggerAvailability.Unavailable;

        val msvcBinaryPath = "vsdbg.exe";

        val msvcFile = msvcPath().resolve(msvcBinaryPath);
        if (!Files.exists(msvcFile))
            return DebuggerAvailability.NeedToDownload;

        val msvcUrl = msvcUrl();
        if (msvcUrl == null) //Fallback in case falsepattern.com goes down
            return new DebuggerAvailability.MSVCBinaries(msvcFile);

        val versions = loadDebuggerVersions(DebuggerKind.MSVC);

        if (!Objects.equals(versions.get(MSVC_PROPERTY_NAME), msvcUrl.version))
            return DebuggerAvailability.NeedToUpdate;

        return new DebuggerAvailability.MSVCBinaries(msvcFile);
    }

    @SneakyThrows
    public DownloadResult downloadDebugger(@Nullable Project project, DebuggerKind debuggerKind) {
        val result = ProgressManager.getInstance()
                                    .runProcessWithProgressSynchronously(() -> downloadDebuggerSynchronously(debuggerKind),
                                                                         ZigBundle.message("dialog.title.download.debugger"),
                                                                         true,
                                                                         project);

        if (result instanceof DownloadResult.Ok) {
            Notifications.Bus.notify(new Notification(
                    "ZigBrains.Debugger.Warn",
                    ZigBundle.message("notification.title.debugger"),
                    ZigBundle.message("notification.content.debugger.successfully.downloaded"),
                    NotificationType.INFORMATION
            ));
        } else if (result instanceof DownloadResult.Failed) {
            Notifications.Bus.notify(new Notification(
                    "ZigBrains.Project",
                    ZigBundle.message("notification.title.debugger"),
                    ZigBundle.message("notification.content.debugger.downloading.failed"),
                    NotificationType.ERROR
            ));
        }

        return result;
    }

    private DownloadResult downloadDebuggerSynchronously(DebuggerKind kind) {
        val baseDir = basePath(kind);

        val downloadableBinaries = switch (kind) {
            case LLDB -> {
                val urls = lldbUrls();
                if (urls == null)
                    yield null;

                val fwUrl = urls.framework.toString();
                val feUrl = urls.frontend.toString();
                yield List.of(
                        new DownloadableDebuggerBinary(fwUrl, LLDB_FRAMEWORK_PROPERTY_NAME, fileNameWithoutExtension(fwUrl)),
                        new DownloadableDebuggerBinary(feUrl, LLDB_FRONTEND_PROPERTY_NAME, fileNameWithoutExtension(feUrl))
                             );
            }
            case GDB -> {
                val gdbUrl = gdbUrl();
                if (gdbUrl == null)
                    yield null;
                val url = gdbUrl.toString();
                yield List.of(new DownloadableDebuggerBinary(url, GDB_PROPERTY_NAME, fileNameWithoutExtension(url)));
            }
            case MSVC -> {
                val msvcUrl = msvcUrl();
                if (msvcUrl == null)
                    yield null;

                AtomicReference<Boolean> accepted = new AtomicReference<>(false);

                ApplicationManager.getApplication().invokeAndWait(() -> {
                    val dialog = new DialogBuilder();
                    dialog.setTitle(msvcUrl.dialogTitle);
                    dialog.addCancelAction().setText("Reject");
                    dialog.addOkAction().setText("Accept");
                    val centerPanel = new JBPanel<>();
                    val hyperlink = new HyperlinkLabel();
                    hyperlink.setTextWithHyperlink(msvcUrl.dialogBody);
                    hyperlink.setHyperlinkTarget(msvcUrl.dialogLink);
                    hyperlink.addHyperlinkListener(new BrowserHyperlinkListener());
                    centerPanel.add(hyperlink);
                    dialog.centerPanel(centerPanel);
                    accepted.set(dialog.showAndGet());
                });

                if (!accepted.get())
                    yield null;

                yield List.of(new DownloadableDebuggerBinary(msvcUrl.url, MSVC_PROPERTY_NAME, msvcUrl.version, "extension/debugAdapters/vsdbg/bin"));
            }
        };

        if (downloadableBinaries == null)
            return DownloadResult.NoUrls;

        try {
            downloadAndUnarchive(baseDir, downloadableBinaries);
            return new DownloadResult.Ok(baseDir);
        } catch (IOException e) {
            //TODO logging
            e.printStackTrace();
            return new DownloadResult.Failed(e.getMessage());
        }
    }

    private void downloadAndUnarchive(Path baseDir, List<DownloadableDebuggerBinary> binariesToDownload)
            throws IOException {
        val service = DownloadableFileService.getInstance();

        val downloadDir = baseDir.toFile();
        FileUtil.deleteRecursively(baseDir);

        val descriptions = binariesToDownload.stream().map(it -> service.createFileDescription(it.url, fileName(it.url))).toList();

        val downloader = service.createDownloader(descriptions, "Debugger downloading");
        val downloadDirectory = downloadPath().toFile();
        val downloadResults = downloader.download(downloadDirectory);

        val versions = new Properties();
        for (val result: downloadResults) {
            val downloadUrl = result.getSecond().getDownloadUrl();
            val binaryToDownload = binariesToDownload.stream()
                                                     .filter(it -> Objects.equals(it.url, downloadUrl))
                                                     .findFirst()
                                                     .orElseThrow(() -> new IOException("Failed to find matching download URL!"));
            val propertyName = binaryToDownload.propertyName;
            val archiveFile = result.getFirst();
            Unarchiver.unarchive(archiveFile, downloadDir, binaryToDownload.prefix);
            archiveFile.delete();
            versions.put(propertyName, binaryToDownload.version);
        }

        saveVersionsFile(baseDir, versions);
    }

    public Properties loadDebuggerVersions(DebuggerKind kind) {
        return loadVersions(basePath(kind));
    }

    public void saveDebuggerVersions(DebuggerKind kind, Properties versions) {
        saveVersionsFile(basePath(kind), versions);
    }

    private void saveVersionsFile(Path basePath, Properties versions) {
        val file = basePath.resolve(DEBUGGER_VERSIONS);
        try (val writer = Files.newBufferedWriter(file)){
            versions.store(writer, "");
        } catch (IOException e) {
            //TODO logging
            e.printStackTrace();
        }
    }

    private Properties loadVersions(Path basePath) {
        val versions = new Properties();
        val versionsFile = basePath.resolve(DEBUGGER_VERSIONS);

        if (Files.exists(versionsFile)) {
            try(val reader = Files.newBufferedReader(versionsFile)) {
                versions.load(reader);
            } catch (IOException e) {
                //TODO logging
                e.printStackTrace();
            }
        }
        return versions;
    }

    private Path basePath(DebuggerKind kind) {
        return switch (kind) {
            case LLDB -> lldbPath();
            case GDB -> gdbPath();
            case MSVC -> msvcPath();
        };
    }

    static Path downloadPath() {
        return Paths.get(PathManager.getTempPath());
    }

    private static Path lldbPath() {
        return ZigPathManager.pluginDirInSystem().resolve("lldb");
    }

    private static Path gdbPath() {
        return ZigPathManager.pluginDirInSystem().resolve("gdb");
    }

    private static Path msvcPath() {
        return ZigPathManager.pluginDirInSystem().resolve("msvc");
    }

    private LLDBUrls lldbUrls() {
        val lldb = UrlProvider.lldb(OS.CURRENT, CpuArch.CURRENT);
        if (lldb == null)
            return null;

        val lldbFrontend = UrlProvider.lldbFrontend(OS.CURRENT, CpuArch.CURRENT);
        if (lldbFrontend == null)
            return null;

        return new LLDBUrls(lldb, lldbFrontend);
    }

    private record LLDBUrls(URL framework, URL frontend) {}

    private URL gdbUrl() {
        return UrlProvider.gdb(OS.CURRENT, CpuArch.CURRENT);
    }

    private MSVCUrl msvcUrl() {
        String dlKey = switch (CpuArch.CURRENT) {
            case X86 -> "downloadX86";
            case X86_64 -> "downloadX86_64";
            case ARM64 -> "downloadARM64";
            default -> null;
        };
        if (dlKey == null)
            return null;

        val props = msvcProvider.msvcProperties();
        val version = props.getProperty("version");
        val url = props.getProperty(dlKey);
        if (url == null || version == null)
            return null;

        return new MSVCUrl(url, version, props.getProperty("dialogTitle"), props.getProperty("dialogBody"), props.getProperty("dialogLink"));
    }

    private record MSVCUrl(String url, String version, String dialogTitle, String dialogBody, String dialogLink) {}

    private static String fileName(String url) {
        return url.substring(url.lastIndexOf('/') + 1);
    }

    private static String fileNameWithoutExtension(String url) {
        url = fileName(url);
        url = removeSuffix(url, ".zip");
        url = removeSuffix(url, ".tar.gz");
        return url;
    }

    private static String removeSuffix(String str, String suffix) {
        if (str.endsWith(suffix))
            str = str.substring(0, str.length() - suffix.length());
        return str;
    }
    private static final String DEBUGGER_VERSIONS = "versions.properties";
    private static final String LLDB_FRONTEND_PROPERTY_NAME = "lldbFrontend";
    private static final String LLDB_FRAMEWORK_PROPERTY_NAME = "lldbFramework";
    private static final String GDB_PROPERTY_NAME = "gdb";
    private static final String MSVC_PROPERTY_NAME = "msvc";

    public sealed interface DownloadResult {
        record Ok(Path baseDir) implements DownloadResult {}
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        final class NoUrls implements DownloadResult {}
        NoUrls NoUrls = new NoUrls();
        record Failed(String message) implements DownloadResult {}
    }

    private enum Unarchiver {
        ZIP {
            @Override
            protected String extension() {
                return "zip";
            }

            @Override
            protected Decompressor createDecompressor(File file) {
                return new Decompressor.Zip(file);
            }
        },
        TAR {
            @Override
            protected String extension() {
                return "tar.gz";
            }

            @Override
            protected Decompressor createDecompressor(File file) {
                return new Decompressor.Tar(file);
            }
        },
        VSIX {
            @Override
            protected String extension() {
                return "vsix";
            }

            @Override
            protected Decompressor createDecompressor(File file) {
                return new Decompressor.Zip(file);
            }
        };

        protected abstract String extension();
        protected abstract Decompressor createDecompressor(File file);

        static void unarchive(File archivePath, File dst, String prefix) throws IOException {
            val unarchiver = Arrays.stream(values())
                                   .filter(it -> archivePath.getName().endsWith(it.extension()))
                                   .findFirst()
                                   .orElseThrow(() -> new IOException("Failed to find decompressor for file " + archivePath.getName()));
            val dec = unarchiver.createDecompressor(archivePath);
            if (prefix != null) {
                dec.removePrefixPath(prefix);
            }
            dec.extract(dst);
        }
    }

    private record DownloadableDebuggerBinary(String url, String propertyName, String version, String prefix) {
        public DownloadableDebuggerBinary(String url, String propertyName, String version) {
            this(url, propertyName, version, null);
        }
    }
}

File: modules/cidr/src/main/kotlin/com/falsepattern/zigbrains/debugger/toolchain/ZigDebuggerToolchainService.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2024 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.debugger.toolchain

import com.falsepattern.zigbrains.debugger.ZigDebugBundle
import com.falsepattern.zigbrains.shared.coroutine.runInterruptibleEDT
import com.falsepattern.zigbrains.shared.coroutine.runModalOrBlocking
import com.intellij.execution.ExecutionModes.ModalProgressMode
import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.openapi.application.EDT
import com.intellij.openapi.application.PathManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.progress.coroutineToIndicator
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogBuilder
import com.intellij.openapi.util.SystemInfo
import com.intellij.openapi.util.io.toNioPathOrNull
import com.intellij.platform.ide.progress.ModalTaskOwner
import com.intellij.platform.ide.progress.TaskCancellation
import com.intellij.platform.ide.progress.withModalProgress
import com.intellij.ui.BrowserHyperlinkListener
import com.intellij.ui.HyperlinkLabel
import com.intellij.ui.components.JBPanel
import com.intellij.util.application
import com.intellij.util.concurrency.annotations.RequiresEdt
import com.intellij.util.download.DownloadableFileService
import com.intellij.util.io.Decompressor
import com.intellij.util.system.CpuArch
import com.intellij.util.system.OS
import com.jetbrains.cidr.execution.debugger.CidrDebuggerPathManager
import com.jetbrains.cidr.execution.debugger.backend.bin.UrlProvider
import com.jetbrains.cidr.execution.debugger.backend.lldb.LLDBDriverConfiguration
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runInterruptible
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException
import java.net.URL
import java.nio.file.Path
import java.util.Properties
import kotlin.io.path.name
import kotlin.io.path.notExists

@Service
class ZigDebuggerToolchainService {
    suspend fun debuggerAvailability(kind: DebuggerKind): DebuggerAvailability<*> {
        return when(kind) {
            DebuggerKind.LLDB -> lldbAvailability()
            DebuggerKind.GDB -> gdbAvailability()
            DebuggerKind.MSVC -> msvcAvailability()
        }
    }

    fun lldbAvailability(): DebuggerAvailability<LLDBBinaries> {
//        if (LLDBDriverConfiguration.hasBundledLLDB()) return DebuggerAvailability.Bundled

        val (frameworkPath, frontendPath) = when {
            SystemInfo.isMac -> "LLDB.framework" to "LLDBFrontend"
            SystemInfo.isUnix -> "lib/liblldb.so" to "bin/LLDBFrontend"
            SystemInfo.isWindows -> "bin/liblldb.dll" to "bin/LLDBFrontend.exe"
            else -> return DebuggerAvailability.Unavailable
        }

        val lldbPath = lldbPath()
        val frameworkFile = lldbPath.resolve(frameworkPath)
        val frontendFile = lldbPath.resolve(frontendPath)
        if (frameworkFile.notExists() || frontendFile.notExists()) return DebuggerAvailability.NeedToDownload

        val versions = loadDebuggerVersions(DebuggerKind.LLDB)
        val (lldbFrameworkUrl, lldbFrontendUrl) = lldbUrls() ?: return DebuggerAvailability.Unavailable

        val lldbFrameworkVersion = fileNameWithoutExtension(lldbFrameworkUrl.toString())
        val lldbFrontendVersion = fileNameWithoutExtension(lldbFrontendUrl.toString())

        if (versions[LLDB_FRAMEWORK_PROPERTY_NAME] != lldbFrameworkVersion ||
            versions[LLDB_FRONTEND_PROPERTY_NAME] != lldbFrontendVersion) return DebuggerAvailability.NeedToUpdate

        return DebuggerAvailability.Binaries(LLDBBinaries(frameworkFile, frontendFile))
    }

    fun gdbAvailability(): DebuggerAvailability<GDBBinaries> {
        if (SystemInfo.isMac) return DebuggerAvailability.Unavailable
//        if (CidrDebuggerPathManager.getBundledGDBBinary().exists()) return DebuggerAvailability.Bundled

        val gdbBinaryPath = when {
            SystemInfo.isUnix -> "bin/gdb"
            SystemInfo.isWindows -> "bin/gdb.exe"
            else -> return DebuggerAvailability.Unavailable
        }

        val gdbFile = gdbPath().resolve(gdbBinaryPath)
        if (gdbFile.notExists()) return DebuggerAvailability.NeedToDownload

        val versions = loadDebuggerVersions(DebuggerKind.GDB)
        val gdbUrl = gdbUrl() ?: return DebuggerAvailability.Unavailable

        val gdbVersion = fileNameWithoutExtension(gdbUrl.toString())

        if (versions[GDB_PROPERTY_NAME] != gdbVersion) return DebuggerAvailability.NeedToUpdate

        return DebuggerAvailability.Binaries(GDBBinaries(gdbFile))
    }

    suspend fun msvcAvailability(): DebuggerAvailability<MSVCBinaries> {
//        if (!SystemInfo.isWindows) return DebuggerAvailability.Unavailable

        val msvcBinaryPath = "vsdbg.exe"

        val msvcFile = msvcPath().resolve(msvcBinaryPath)
        if (msvcFile.notExists()) return DebuggerAvailability.NeedToDownload

        val msvcUrl = msvcUrl() ?: return DebuggerAvailability.Binaries(MSVCBinaries(msvcFile))

        val versions = loadDebuggerVersions(DebuggerKind.MSVC)

        if (versions[MSVC_PROPERTY_NAME] != msvcUrl.version) return DebuggerAvailability.NeedToUpdate

        return DebuggerAvailability.Binaries(MSVCBinaries(msvcFile))
    }

    @RequiresEdt
    private suspend fun doDownloadDebugger(project: Project? = null, debuggerKind: DebuggerKind): DownloadResult {
        val baseDir = debuggerKind.basePath()
        val downloadableBinaries = when(debuggerKind) {
            DebuggerKind.LLDB -> {
                val (lldbFrameworkUrl, lldbFrontendUrl) = lldbUrls()?.run { first.toString() to second.toString() } ?: return DownloadResult.NoUrls
                listOf(
                    DownloadableDebuggerBinary(lldbFrameworkUrl, LLDB_FRAMEWORK_PROPERTY_NAME, fileNameWithoutExtension(lldbFrameworkUrl)),
                    DownloadableDebuggerBinary(lldbFrontendUrl, LLDB_FRONTEND_PROPERTY_NAME, fileNameWithoutExtension(lldbFrontendUrl))
                )
            }
            DebuggerKind.GDB -> {
                val gdbUrl = gdbUrl()?.run { toString() } ?: return DownloadResult.NoUrls
                listOf(DownloadableDebuggerBinary(gdbUrl, GDB_PROPERTY_NAME, fileNameWithoutExtension(gdbUrl)))
            }
            DebuggerKind.MSVC -> {
                val msvcUrl = msvcUrl() ?: return DownloadResult.NoUrls

                val dialog = DialogBuilder()
                dialog.setTitle(msvcUrl.dialogTitle)
                dialog.addCancelAction().setText("Reject")
                dialog.addOkAction().setText("Accept")
                val centerPanel = JBPanel<JBPanel<*>>()
                val hyperlink = HyperlinkLabel()
                hyperlink.setTextWithHyperlink(msvcUrl.dialogBody)
                hyperlink.setHyperlinkText(msvcUrl.dialogLink)
                hyperlink.addHyperlinkListener(BrowserHyperlinkListener())
                centerPanel.add(hyperlink)
                dialog.centerPanel(centerPanel)
                if (!dialog.showAndGet()) return DownloadResult.NoUrls

                listOf(DownloadableDebuggerBinary(msvcUrl.url, MSVC_PROPERTY_NAME, msvcUrl.version, "extension/debugAdapters/vsdbg/bin"))
            }
        }

        try {
            downloadAndUnArchive(baseDir, downloadableBinaries)
            return DownloadResult.Ok(baseDir)
        } catch (e: IOException) {
            //TODO logging
            e.printStackTrace()
            return DownloadResult.Failed(e.message)
        }
    }

    @RequiresEdt
    suspend fun downloadDebugger(project: Project? = null, debuggerKind: DebuggerKind): DownloadResult {
        val result = doDownloadDebugger(project, debuggerKind)

        when(result) {
            is DownloadResult.Ok -> {
                Notification(
                    "zigbrains",
                    ZigDebugBundle.message("notification.title.debugger"),
                    ZigDebugBundle.message("notification.content.debugger.successfully.downloaded"),
                    NotificationType.INFORMATION
                ).notify(project)
            }
            is DownloadResult.Failed -> {
                Notification(
                    "zigbrains",
                    ZigDebugBundle.message("notification.title.debugger"),
                    ZigDebugBundle.message("notification.content.debugger.downloading.failed"),
                    NotificationType.ERROR
                ).notify(project)
            }
            else -> Unit
        }

        return result
    }

    @Throws(IOException::class)
    @RequiresEdt
    private suspend fun downloadAndUnArchive(baseDir: Path, binariesToDownload: List<DownloadableDebuggerBinary>) {
        val service = DownloadableFileService.getInstance()

        val downloadDir = baseDir.toFile()
        downloadDir.deleteRecursively()

        val descriptions = binariesToDownload.map {
            service.createFileDescription(it.url, fileName(it.url))
        }

        val downloader = service.createDownloader(descriptions, "Debugger downloading")
        val downloadDirectory = downloadPath().toFile()
        val downloadResults = withContext(Dispatchers.IO) {
            coroutineToIndicator {
                downloader.download(downloadDirectory)
            }
        }
        val versions = Properties()
        for (result in downloadResults) {
            val downloadUrl = result.second.downloadUrl
            val binaryToDownload = binariesToDownload.first { it.url == downloadUrl }
            val propertyName = binaryToDownload.propertyName
            val archiveFile = result.first
            Unarchiver.unarchive(archiveFile, downloadDir)
            archiveFile.delete()
            versions[propertyName] = binaryToDownload.version
        }

        saveVersionsFile(baseDir, versions)
    }

    private fun lldbUrls(): Pair<URL, URL>? {
        val lldb = UrlProvider.lldb(OS.CURRENT, CpuArch.CURRENT) ?: return null
        val lldbFrontend = UrlProvider.lldbFrontend(OS.CURRENT, CpuArch.CURRENT) ?: return null
        return lldb to lldbFrontend
    }

    private fun gdbUrl(): URL? = UrlProvider.gdb(OS.CURRENT, CpuArch.CURRENT)

    private suspend fun msvcUrl(): MSVCUrl? {
        val dlKey = when(CpuArch.CURRENT) {
            CpuArch.X86 -> "downloadX86"
            CpuArch.X86_64 -> "downloadX86_64"
            CpuArch.ARM64 -> "downloadARM64"
            else -> return null
        }

        val props = msvcMetadata()
        val version = props.getProperty("version") ?: return null
        val url = props.getProperty(dlKey) ?: return null
        return MSVCUrl(url, version, props.getProperty("dialogTitle")!!, props.getProperty("dialogBody")!!, props.getProperty("dialogLink")!!)
    }

    private data class MSVCUrl(
        val url: String,
        val version: String,
        val dialogTitle: String,
        val dialogBody: String,
        val dialogLink: String
    )

    private fun loadDebuggerVersions(kind: DebuggerKind): Properties = loadVersions(kind.basePath())

    private fun saveVersionsFile(basePath: Path, versions: Properties) {
        val file = basePath.resolve(DEBUGGER_VERSIONS).toFile()
        try {
            versions.store(file.bufferedWriter(), "")
        } catch (e: IOException) {
            LOG.warn("Failed to save `${basePath.name}/${file.name}`", e)
        }
    }

    private fun loadVersions(basePath: Path): Properties {
        val versions = Properties()
        val versionsFile = basePath.resolve(DEBUGGER_VERSIONS).toFile()

        if (versionsFile.exists()) {
            try {
                versionsFile.bufferedReader().use { versions.load(it) }
            } catch (e: IOException) {
                LOG.warn("Failed to load `${basePath.name}/${versionsFile.name}`", e)
            }
        }

        return versions
    }

    private fun DebuggerKind.basePath(): Path {
        return when(this) {
            DebuggerKind.LLDB -> lldbPath()
            DebuggerKind.GDB -> gdbPath()
            DebuggerKind.MSVC -> msvcPath()
        }
    }

    companion object {
        private val LOG = logger<ZigDebuggerToolchainService>()

        private const val DEBUGGER_VERSIONS: String = "versions.properties"

        private const val LLDB_FRONTEND_PROPERTY_NAME = "lldbFrontend"
        private const val LLDB_FRAMEWORK_PROPERTY_NAME = "lldbFramework"
        private const val GDB_PROPERTY_NAME = "gdb"
        private const val MSVC_PROPERTY_NAME = "msvc"

        fun downloadPath() = tempPluginDir
        private fun lldbPath() = pluginDir.resolve("lldb")
        private fun gdbPath() = pluginDir.resolve("gdb")
        private fun msvcPath() = pluginDir.resolve("msvc")

        private val pluginDir get() = PathManager.getSystemDir().resolve("zigbrains")

        private val tempPluginDir get(): Path = PathManager.getTempPath().toNioPathOrNull()!!.resolve("zigbrains")

        private fun fileNameWithoutExtension(url: String): String {
            return url.substringAfterLast("/").removeSuffix(".zip").removeSuffix(".tar.gz")
        }

        private fun fileName(url: String): String {
            return url.substringAfterLast("/")
        }
    }

    private enum class Unarchiver {
        ZIP {
            override val extension = "zip"
            override fun createDecompressor(file: File) = Decompressor.Zip(file)
        },
        TAR {
            override val extension = "tar.gz"
            override fun createDecompressor(file: File) = Decompressor.Tar(file)
        },
        VSIX {
            override val extension = "vsix"
            override fun createDecompressor(file: File) = Decompressor.Zip(file)
        };

        protected abstract val extension: String
        protected abstract fun createDecompressor(file: File): Decompressor

        companion object {
            @Throws(IOException::class)
            suspend fun unarchive(archivePath: File, dst: File, prefix: String? = null) {
                runInterruptible {
                    val unarchiver = entries.find { archivePath.name.endsWith(it.extension) } ?: error("Unexpected archive type: $archivePath")
                    val dec = unarchiver.createDecompressor(archivePath)
                    if (prefix != null) {
                        dec.removePrefixPath(prefix)
                    }
                    dec.extract(dst)
                }
            }
        }
    }

    sealed class DownloadResult {
        class Ok(val baseDir: Path): DownloadResult()
        data object NoUrls: DownloadResult()
        class Failed(val message: String?): DownloadResult()
    }

    @JvmRecord
    private data class DownloadableDebuggerBinary(val url: String, val propertyName: String, val version: String, val prefix: String? = null)
}

val zigDebuggerToolchainService get() = application.service<ZigDebuggerToolchainService>()
